File: LinkedLists/2095.Delete the Middle Node of a Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* deleteMiddle(ListNode* head) 
    {
        ListNode* p=head;
        ListNode* q=NULL;
        if(p->next==NULL)
        {
            return q;
        }
        if(p->next->next==NULL)
        {
            p->next=NULL;
            return p;
        }
        int count=0;
        q=head;
        ListNode* r=head;
        while(p!=NULL && p->next!=NULL)
        {
            count+=1;
            p=p->next->next;
            q=q->next;
        }
        for(int i=0;i<count-1;i++)
        {
            r=r->next;
        }
        r->next=q->next;
        return head;
        
        
//         p=head->next->next;
//         q=head;
//         while(p!=NULL && p->next!=NULL)
//         {
//             p=p->next->next;
//             q=q->next;
//         }
//         q->next=q->next->next;
//         return head;
    }
};

File: Trees/2095 Delete the Middle Node of a Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def deleteMiddle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head.next==None:
            return None 
        if head.next.next==None:
            head.next=None
            return head
        p=head.next.next
        q=head
        while p and p.next:
            p = p.next.next
            q = q.next
        q.next=q.next.next
        return head
